lstchain/calib/camera/sampling_interval_coefficient_calculate.py

```diff
@@ -0,0 +1,170 @@
+    """
+    def __init__(self):
+        self.peak_count = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+        self.fc_count = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+
+        self.peak_count_stuck ={}
+        self.fc_count_stuck = {}
+
+        self.sampling_interval_coefficient = {}
+        self.charge_array_after_corr ={}
+        self.charge_reso_array_after_corr ={}
+
+        self.charge_reso_final = np.zeros(N_PIXELS)
+        self.used_run = np.zeros(N_PIXELS)
+        self.sampling_interval_coefficient_final = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+
+    def increment_peak_count(self, event, tel_id, gain, r0_r1_calibrator):
+
+        waveform = event.r1.tel[tel_id].waveform
+        
+        first_capacitors = r0_r1_calibrator.first_cap
+        r1_sample_start = r0_r1_calibrator.r1_sample_start.tel[tel_id]
+
+        # Check pulse
+        pulse_event_flag = np.sum(np.max(waveform[gain], axis=1) > 100) > 1800
+    
+        if pulse_event_flag:
+
+            # find pulse peak position
+            pulse_peak = np.argmax(waveform[gain], axis=1)
+            fc = first_capacitors[tel_id][gain]
+            pulse_peak_abs_pos = (fc + r1_sample_start + pulse_peak) % N_CAPACITORS_CHANNEL
+            
+            # increment peak count array
+            self.peak_count[np.arange(N_PIXELS), pulse_peak_abs_pos] += 1
+            self.fc_count[np.arange(N_PIXELS), fc % N_CAPACITORS_CHANNEL] += 1
+            
+
+    def stuck_single_sampling_interval(self, file_list, gain):
+
+        for single_file in file_list:
+
+            run_id, gain_in_file, peak_count, fc_count = load_single_fits_sampling_interval(single_file)
+            
+            if gain_in_file == gain:
+                if not run_id in self.peak_count_stuck.keys():
+                    self.peak_count_stuck[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+                    self.fc_count_stuck[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL])
+                
+                self.peak_count_stuck[run_id] += peak_count
+                self.fc_count_stuck[run_id] += fc_count
+
+
+    def convert_to_samp_interval_coefficient(self, gain):
+        # convert peak counts to sampling interval coefficient
+
+        for run_id in self.peak_count_stuck.keys():
+            self.sampling_interval_coefficient[run_id] = np.zeros([N_PIXELS, N_CAPACITORS_CHANNEL + N_SAMPLES])
+
+            for pixel in range(N_PIXELS):
+                self.sampling_interval_coefficient[run_id][pixel, :N_CAPACITORS_CHANNEL] = \
+                    self.peak_count_stuck[run_id][pixel] / np.sum(self.peak_count_stuck[run_id][pixel]) * N_CAPACITORS_CHANNEL
+                
+                self.sampling_interval_coefficient[run_id][pixel, N_CAPACITORS_CHANNEL:] = \
+                    self.sampling_interval_coefficient[run_id][pixel, :N_SAMPLES]
+
+                
+    def set_charge_array(self, gain):
+        self.charge_array_before_corr = np.zeros([N_PIXELS, 60000])
```

Comment:
Why 60000?

Reply:
I modified related codes. If `max_events` is given, this given value will be used to generate a new array.
If not (`max_events` is `None` by default), just use a large enough number (55000) to cover all events in a single subrun.
(A single subrun usually contains ~53000 events.)
I'm sure it is not the best way... so I'm very happy if you have any good idea...

Reply:
You can actually get the number of events by doing `len(source.multi_file)`, I will also add `__len__` to the `LSTEventSource`, but that will require a new release.

A more flexible solution would be to use a python list and append to it, converting to a 2d numpy array at the end.